StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/Helpers/FixAllContextHelper.cs

```diff
@@ -0,0 +1,267 @@
+
+            var diagnosticAnalyzerType = typeof(DiagnosticAnalyzer);
+
+            var analyzers = ImmutableDictionary.CreateBuilder<string, ImmutableArray<Type>>();
+
+            foreach (var type in assembly.DefinedTypes)
+            {
+                if (type.IsSubclassOf(diagnosticAnalyzerType) && !type.IsAbstract)
+                {
+                    Type analyzerType = type.AsType();
+                    DiagnosticAnalyzer analyzer = (DiagnosticAnalyzer)Activator.CreateInstance(analyzerType);
+                    foreach (var descriptor in analyzer.SupportedDiagnostics)
+                    {
+                        ImmutableArray<Type> types;
+                        if (analyzers.TryGetValue(descriptor.Id, out types))
+                        {
+                            types = types.Add(analyzerType);
+                        }
+                        else
+                        {
+                            types = ImmutableArray.Create(analyzerType);
+                        }
+
+                        analyzers[descriptor.Id] = types;
+                    }
+                }
+            }
+
+            return analyzers.ToImmutable();
+        }
+
+        private static ImmutableArray<DiagnosticAnalyzer> GetDiagnosticAnalyzersForContext(FixAllContext fixAllContext)
+        {
+            return DiagnosticAnalyzers
+                .Where(x => fixAllContext.DiagnosticIds.Contains(x.Key))
+                .SelectMany(x => x.Value)
+                .Distinct()
+                .Select(type => (DiagnosticAnalyzer)Activator.CreateInstance(type))
+                .ToImmutableArray();
+        }
+
+        private static async Task<ImmutableArray<Diagnostic>> GetAllDiagnosticsAsync(FixAllContext fixAllContext, Project project, ImmutableArray<DiagnosticAnalyzer> analyzers)
+        {
+            bool includeCompilerDiagnostics = fixAllContext.DiagnosticIds.Any(x => x.StartsWith("CS", StringComparison.Ordinal));
+            var compilation = await project.GetCompilationAsync(fixAllContext.CancellationToken).ConfigureAwait(false);
+
+            var diagnostics = ImmutableArray<Diagnostic>.Empty;
+            if (analyzers.Any())
+            {
+                var compilationWithAnalyzers = compilation.WithAnalyzers(analyzers, project.AnalyzerOptions, fixAllContext.CancellationToken);
+                if (GetAnalyzerSyntaxDiagnosticsAsync != null && GetAnalyzerSemanticDiagnosticsAsync != null)
+                {
+                    // This whole block is workaround code for issues with Roslyn 1.1...
+                    compilationWithAnalyzers.Compilation.GetDeclarationDiagnostics(fixAllContext.CancellationToken);
+
+                    var tasks = new List<Task<ImmutableArray<Diagnostic>>>();
+                    foreach (var document in project.Documents)
+                    {
+                        var syntaxTree = await document.GetSyntaxTreeAsync(fixAllContext.CancellationToken).ConfigureAwait(false);
```

Comment:
:memo: The processing of documents cannot run in parallel, because a single `CompilationWithAnalyzers` instance is used for each of the documents. We are using a single instance to ensure that all diagnostics are identified following a *single* call to `GetDeclarationDiagnostics`, as opposed to calling it once per document.